// @unblu/floating-js-api/dist/lib/shared/internal/module/base-general-module.js
import { InternalModule } from "./module";
export var BaseGeneralFunction;
(function (BaseGeneralFunction) {
    BaseGeneralFunction["startConversation"] = "startConversation";
    BaseGeneralFunction["setNewConversationInterceptor"] = "setNewConversationInterceptor";
    BaseGeneralFunction["joinConversation"] = "joinConversation";
    BaseGeneralFunction["openConversation"] = "openConversation";
    BaseGeneralFunction["getActiveConversation"] = "getActiveConversation";
    BaseGeneralFunction["getConversations"] = "getConversations";
    BaseGeneralFunction["getNotificationCount"] = "getNotificationCount";
    BaseGeneralFunction["getPersonInfo"] = "getPersonInfo";
    BaseGeneralFunction["setPersonNickname"] = "setPersonNickname";
    BaseGeneralFunction["addPersonLabel"] = "addPersonLabel";
    BaseGeneralFunction["removePersonLabel"] = "removePersonLabel";
})(BaseGeneralFunction || (BaseGeneralFunction = {}));
export class BaseGeneralModule extends InternalModule {
    constructor(bridge) {
        super(bridge, 'general');
    }
    startConversation(type, visitorName, visitorData, recipient) {
        return this.callApiFunction(BaseGeneralFunction.startConversation, [type, visitorName, visitorData, recipient]);
    }
    setNewConversationInterceptor(callback) {
        return this.callApiFunction(BaseGeneralFunction.setNewConversationInterceptor, [callback]);
    }
    joinConversation(pin, visitorName) {
        return this.callApiFunction(BaseGeneralFunction.joinConversation, [pin, visitorName]);
    }
    openConversation(conversationId) {
        return this.callApiFunction(BaseGeneralFunction.openConversation, [conversationId]);
    }
    getActiveConversation() {
        return this.callApiFunction(BaseGeneralFunction.getActiveConversation, []);
    }
    getConversations() {
        return this.callApiFunction(BaseGeneralFunction.getConversations, []);
    }
    getNotificationCount() {
        return this.callApiFunction(BaseGeneralFunction.getNotificationCount, []);
    }
    getPersonInfo() {
        return this.callApiFunction(BaseGeneralFunction.getPersonInfo, []);
    }
    setPersonNickname(nickname) {
        return this.callApiFunction(BaseGeneralFunction.setPersonNickname, [nickname]);
    }
    addPersonLabel(labelName) {
        return this.callApiFunction(BaseGeneralFunction.addPersonLabel, [labelName]);
    }
    removePersonLabel(labelName) {
        return this.callApiFunction(BaseGeneralFunction.removePersonLabel, [labelName]);
    }
}
