// @unblu/floating-js-api/dist/lib/shared/internal/module/conversation-module.js
import { InternalModule } from './module';
export class ConversationModule extends InternalModule {
    constructor(bridge) {
        super(bridge, 'conversation');
    }
    getConnectionState(conversationId) {
        return this.callApiFunction("getConnectionState" /* ConversationFunction.getConnectionState */, [conversationId]);
    }
    getConversationState(conversationId) {
        return this.callApiFunction("getConversationState" /* ConversationFunction.getConversationState */, [conversationId]);
    }
    getCallState(conversationId) {
        return this.callApiFunction("getCallState" /* ConversationFunction.getCallState */, [conversationId]);
    }
    startAudioCall(conversationId) {
        return this.callApiFunction("startAudioCall" /* ConversationFunction.startAudioCall */, [conversationId]);
    }
    startVideoCall(conversationId) {
        return this.callApiFunction("startVideoCall" /* ConversationFunction.startVideoCall */, [conversationId]);
    }
    endConversation(conversationId) {
        return this.callApiFunction("endConversation" /* ConversationFunction.endConversation */, [conversationId]);
    }
    leaveConversation(conversationId) {
        return this.callApiFunction("leaveConversation" /* ConversationFunction.leaveConversation */, [conversationId]);
    }
    closeConversation(conversationId) {
        return this.callApiFunction("closeConversation" /* ConversationFunction.closeConversation */, [conversationId]);
    }
    async setVisitorData(visitorData) {
        return this.callApiFunction("setVisitorData" /* ConversationFunction.setVisitorData */, [visitorData]);
    }
}
