// @unblu/floating-js-api/dist/lib/index.js
import { UnbluApi } from './unblu-api';
import { Conversation, } from './shared/conversation';
import { ApiState } from "./shared/api-state";
import { UnbluStaticApi } from './unblu-static-api';
import { CallState } from './shared/model/call-state';
import { ConnectionState } from './shared/model/connection-state';
import { ConversationType } from './shared/model/conversation-type';
import { ConversationRecipientType } from "./shared/model/conversation-recipient-type";
import { NewConversationRecipientType } from "./shared/model/new-conversation-recipient-type";
import { AgentAvailabilityState } from './shared/model/agent-availability-state';
import { ConversationState } from './shared/model/conversation-state';
import { UnbluApiError, UnbluErrorType } from './shared/unblu-api-error';
import { UnbluUiApi } from "./unblu-api-ui";
import { ConversationParticipationType } from "./shared/model/conversation-participation-type";
export { AgentAvailabilityState };
export { ApiState, UnbluStaticApi };
export { CallState };
export { ConnectionState };
export { ConversationState };
export { ConversationType };
export { UnbluErrorType };
export { Conversation };
export { UnbluApi, ConversationRecipientType, NewConversationRecipientType };
export { UnbluApiError };
export { UnbluUiApi };
export { ConversationParticipationType };
export default {
    /**
     * Access to the Unblu Floating API
     */
    api: new UnbluStaticApi(),
    AgentAvailabilityState: AgentAvailabilityState,
    CallState: CallState,
    ConnectionState: ConnectionState,
    ConversationState: ConversationState,
    ConversationType: ConversationType,
    UnbluErrorType: UnbluErrorType
};
