// @unblu/floating-js-api/dist/lib/internal/module/general-module.js
import { BaseGeneralModule } from "../../shared/internal/module/base-general-module";
export class GeneralModule extends BaseGeneralModule {
    toggleIndividualUi() {
        return this.callApiFunction("toggleIndividualUi" /* GeneralFunction.toggleIndividualUi */, []);
    }
    openPinEntryUi() {
        return this.callApiFunction("openPinEntryUi" /* GeneralFunction.openPinEntryUi */, []);
    }
    openOverviewUi() {
        return this.callApiFunction("openOverviewUi" /* GeneralFunction.openOverviewUi */, []);
    }
    popoutIndividualUi() {
        return this.callApiFunction("popoutIndividualUi" /* GeneralFunction.popoutIndividualUi */, []);
    }
    popinIndividualUi() {
        return this.callApiFunction("popinIndividualUi" /* GeneralFunction.popinIndividualUi */, []);
    }
    maximizeIndividualUi() {
        return this.callApiFunction("maximizeIndividualUi" /* GeneralFunction.maximizeIndividualUi */, []);
    }
    minimizeIndividualUi() {
        return this.callApiFunction("minimizeIndividualUi" /* GeneralFunction.minimizeIndividualUi */, []);
    }
    openIndividualUi() {
        return this.callApiFunction("openIndividualUi" /* GeneralFunction.openIndividualUi */, []);
    }
    collapseIndividualUi() {
        return this.callApiFunction("collapseIndividualUi" /* GeneralFunction.collapseIndividualUi */, []);
    }
    getIndividualUiState() {
        return this.callApiFunction("getIndividualUiState" /* GeneralFunction.getIndividualUiState */, []);
    }
    getActiveIndividualUiView() {
        return this.callApiFunction("getActiveIndividualUiView" /* GeneralFunction.getActiveIndividualUiView */, []);
    }
}
