// @unblu/floating-js-api/dist/lib/shared/model/conversation-type.js
export var ConversationType;
(function (ConversationType) {
    /**
     * (Visitor initiated, visitor centered) A visitor creates an incoming chat conversation.
     */
    ConversationType["CHAT_REQUEST"] = "CHAT_REQUEST";
    /**
     * (Visitor initiated, visitor centered) A visitor creates an incoming chat conversation while all agents are offline.
     */
    ConversationType["OFFLINE_CHAT_REQUEST"] = "OFFLINE_CHAT_REQUEST";
    /**
     * (Visitor initiated, visitor centered) A visitor creates an incoming video conversation.
     */
    ConversationType["VIDEO_REQUEST"] = "VIDEO_REQUEST";
    /**
     * (Visitor initiated, visitor centered) A visitor creates an incoming audio conversation.
     */
    ConversationType["AUDIO_REQUEST"] = "AUDIO_REQUEST";
    /**
     * (Visitor initiated, visitor centered) A visitor creates an incoming universal co-browsing conversation.
     */
    ConversationType["HEADLESS_BROWSER_REQUEST"] = "HEADLESS_BROWSER_REQUEST";
    /**
     * (Visitor initiated, visitor centered) A visitor creates an incoming embedded co-browsing conversation.
     */
    ConversationType["DOMCAP_BROWSER_REQUEST"] = "DOMCAP_BROWSER_REQUEST";
    /**
     * (Visitor initiated, visitor centered) A visitor creates an incoming mobile co-browsing conversation.
     */
    ConversationType["MOBILE_COBROWSING_REQUEST"] = "MOBILE_COBROWSING_REQUEST";
    /**
     * (Visitor initiated, visitor centered) A visitor creates an incoming whiteboard conversation.
     */
    ConversationType["WHITEBOARD_REQUEST"] = "WHITEBOARD_REQUEST";
    /**
     * (Visitor initiated, visitor centered) A visitor creates an incoming screen sharing conversation.
     */
    ConversationType["SCREEN_SHARING_REQUEST"] = "SCREEN_SHARING_REQUEST";
    /**
     * (Api initiated, visitor centered) A visitor creates a conversation via API which is not added to the queue and where he can invite other visitor to do
     * embedded
     */
    ConversationType["VISITOR_COBROWSING"] = "VISITOR_COBROWSING";
})(ConversationType || (ConversationType = {}));
