// @unblu/floating-js-api/dist/lib/shared/model/conversation-state.js
export var ConversationState;
(function (ConversationState) {
    /**
     * Initial state of a conversation
     */
    ConversationState["CREATED"] = "CREATED";
    /**
     * Depending of the engagement type of the conversation, the conversation stays in the onboarding state as long as the person in the center of the conversation
     * is doing his onboarding
     */
    ConversationState["ONBOARDING"] = "ONBOARDING";
    /**
     * When a message is send to an unassigned conversation, the conversation first is put into the reboarding state as long as the person in the center of the
     * conversation is doing the reboarding.
     */
    ConversationState["REBOARDING"] = "REBOARDING";
    /**
     * Depending of the engagement type of the conversation, the conversation is put in the queued state when it is inside the queue.
     */
    ConversationState["QUEUED"] = "QUEUED";
    /**
     * Conversation is active
     */
    ConversationState["ACTIVE"] = "ACTIVE";
    /**
     * After there is no assignee of a conversation anymore, the conversation goes to unassigned state until a message from a visitor is written to start reboarding
     * again.
     */
    ConversationState["UNASSIGNED"] = "UNASSIGNED";
    /**
     * Conversation is being ended
     */
    ConversationState["OFFBOARDING"] = "OFFBOARDING";
    /**
     * Conversation is ended
     */
    ConversationState["ENDED"] = "ENDED";
})(ConversationState || (ConversationState = {}));
